### DISCUSSION TITLE
I'm interested more in how to use whatwg/fs with sync access handle

### DISCUSSION THREAD

**Discussion by jimmywarting** [rhashimoto/wa-sqlite#59] [Q&A] [answered]
I plan to run sqlite in a web worker and i also want to use the new whatwg/fs.

I know that whatwg/fs has support for reading/writing files synchronous if you call: `fileHandle.createSyncAccessHandle()`
and it also allows for things such as "write in place" and manually flushing data or what not. 

I was wondering: can i use wa-sqlite in a web worker and using a sync access handle as my preferred database without having some penelty from the asyncify? i still only want the sync part of it all

**Comment by rhashimoto** [comment, marked as answer]
[Edit: This initial reply is mostly wrong.]

I think `createSyncAccessHandle()` has only been implemented in Chromium-based browsers in an origin trial, but yes, if that restriction doesn't bother you then you should be able to write a VFS that works without Asyncify. Note that the origin trial ends in October.

You might want to look at the [example VFS that uses asynchronous access handles](https://github.com/rhashimoto/wa-sqlite/blob/master/src/examples/OriginPrivateFileSystemVFS.js) as a starting point.

**Reply by jimmywarting** [reply]
Nope, it don't bother me, but hey thanks for the suggestion, will have a look at it!

i did notice that the VFS link you sent me actually dose use `createSyncAccessHandle`
then how come that it's so much wrapped in lots of async/await?

**Reply by rhashimoto** [reply]
Oh, right! Sorry, it's been a while since I played with that. Now I'm remembering a little more about it. Apparently `createSyncAccessHandle()` _is_ available in Safari and Chromium-based browsers outside any origin trial.

However, only the reading and writing calls are actually synchronous. Other operations - `truncate()`, `getSize()`, `flush()`, and `close()` - are still asynchronous. This may be a problem in creating a purely synchronous VFS.

**Reply by jimmywarting** [reply]
ah... darn it! you where right. For some reason i thought they where sync. forgot about that. haven't played around much with Sync Access Handle much yet.

had to test it to believe it

```js
async function code() {
  const root = await navigator.storage.getDirectory()
  const fileHandle = await root.getFileHandle('dumdum.db', { create: true })
  const handle = await fileHandle.createSyncAccessHandle()
  console.log(handle.getSize()) // Promise
}

new Worker(URL.createObjectURL(new Blob([code.toString() + ';code()'], { type: 'application/javascript'})), { type: 'module' })
```

now i got a tiny bit disappointed 😞  
now looking at: https://github.com/whatwg/fs/issues/7 which seems to be the blocker

**Reply by rhashimoto** [reply]
The added overhead from Asyncify is often not as bad as people expect. Most database applications are limited by storage performance and not CPU performance. The File System API is pretty slow (slower than IndexedDB can be, actually), even with sync access handle, so the incremental time that Asyncify adds is relatively small.

**Reply by quolpr** [reply]
Yep, I totally agree with @rhashimoto, the overhead is not so big

**Reply by jimmywarting** [reply]
Good to know

**Reply by jimmywarting** [reply]
> The File System API is pretty slow (slower than IndexedDB can be, actually), even with sync access handle

That's pretty bad, how the heck can they design a file system that is slower than using a virtual FS with indexeddb?! can't understand it. i get it that they want to design a secure file system and scanning files and doing atomic writes... it can hurt perfomences so badly depending on what you want to do.

i hope the [new suggestions](https://github.com/whatwg/fs/blob/main/AccessHandle.md) will improve stuff. have long been waiting for the "in-place" option for some other parts of my application

**Reply by rhashimoto** [reply]
The fact that my OPFS VFS was slower than my IndexedDB VFS was initially surprising to me, too, but actually it makes sense. First of all, the OPFS implementations are somewhat new while IndexedDB has been around since 2015 or so, so hopefully we'll see optimizations over time. But also remember that I'm writing filesystems specifically for SQLite, so it's possible that the File System Access API could be faster for other usages.

One reason my OPFS VFS is slower is because I use a non-synchronous access handle for database file reads. The reason I don't use a synchronous access handle instead is to allow multiple connections to share a read lock for concurrent querying.

Another reason my OPFS VFS is slower is there are no write guarantees until a successful flush. This means that any interruption can leave the database file in an inconsistent state so you also need a persistent journal file. The ability to make an atomic write to IndexedDB means that there are several implementation strategies that remove the need to write (and potentially read) a persistent journal file for many/most cases without sacrificing consistency. Using [batch atomic writes](https://github.com/rhashimoto/wa-sqlite/discussions/47) is one approach.

Some IndexedDB implementations allow trading durability for performance, but synchronous access handles do not. Timing comparisons using IndexedDB relaxed durability aren't really apples-to-apples but it is another advantage for IndexedDB if durability isn't a requirement.

You can compare the performance of OPFS [here](https://rhashimoto.github.io/wa-sqlite/demo/opfs.html) (on Safari and Chromium-based browsers) with IndexedDB [here](https://rhashimoto.github.io/wa-sqlite/demo/benchmarks.html).

**Comment by rhashimoto** [comment]
@jimmywarting This discussion (okay, more of a soliloquy) on [changes to the OPFS API](https://github.com/rhashimoto/wa-sqlite/discussions/67) may be of interest to you.

**Reply by jimmywarting** [reply]
Cool, Thanks for sharing.
